### ISSUE TITLE
Suggest syntactic sugar: extension method/function

### ISSUE THREAD

**Issue by gregbacchus** [microsoft/TypeScript#45968]
# Suggestion

Syntactic sugar implementation of extension methods/functions **without** polluting prototypes (or even requiring them at all, i.e. works for interfaces too)

```typescript
// ./my-extension.ts

export extension addDays(this: Date, days: number): Date {
  const result = new Date(this);
  result.setDate(result.getDate() + days);
  return result;
}
```

Can be used:

```typescript
import { addDays } from './my-extension';

const now = new Data();
const nextWeek = now.addDays(7);
```

Simply compiles to something like

```javascript
import { getWeek } from './my-extension';

const now = new Data();
const nextWeek = addDays.bind(now)(7); 
```

So `this` in the extension can be any type/class/interface and it can be used on anything that matches.

```typescript
// GIVEN

interface Person {
  name: string;
}

extension getUppercaseName(this: Person): string {
  // or some really useful function :)
  return this.name.toUppercase();
}

// and

const p1: Person = { name: 'Sarah' };
const p2: { name: 'Jane', age: 34 };

// one could

const n1 = p1.getUppercaseName();
const n2 = p2.getUppercaseName()
const n3 = { name: 'Ali' }.getUppercaseName();
```

## 🔍 Search Terms

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.

* extension
* extension method
* extension function

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Basically it's a simple syntax transformation to make function binding pretty.

Could probably be use for type guards too

```typescript
interface DataResult {
  kind: 'data';
  data: Buffer;
}

interface ErrorResult {
  kind: 'error';
  error: Error;
}

type Result = DataResult | ErrorResult;

extension isError(this: Result): this is ErrorResult {
  return this.kind === 'error';
}
```

## 💻 Use Cases

Any utility library...

**Comment by MartinJohns** [comment]
Duplicate of #9. Used search terms: `extension methods in:title`

> [x] This could be implemented without emitting different JS based on the types of the expressions

I really don't see how this could be implemented without emitting different JS.

**Comment by gregbacchus** [comment]
I don't believe that it is really a duplicate of #9. May be trying to achieve a similar thing. But it's a very simple implementation.

Maybe I don't quite understand:

> This could be implemented without emitting different JS based on the types of the expressions

I don't see how things need to be different based on expression types..?

Are you able to explain?

**Comment by MartinJohns** [comment]
If you don't think it's a duplicate of #9, then it's definitely still a duplicate of #24889. The very exact thing asked.

> I don't see how things need to be different based on expression types..?

You want different JS emitted depending on the declared extension methods. That's generally a no-go in TypeScript, as it goes beyond the type-system scope TypeScript aims at.

**Comment by gregbacchus** [comment]
Thanks for the clarification.

Maybe I didn't explain well. The proposal isn't to emit different JS depending on the types of the expressions. Let me try again:

### Extension method declaration

```typescript
// proposal
extension meExtn(this: MyType): void {
  // ...
}

// treated the same as typescript
function meExtn(this: MyType): void {
  // ...
}

// which emitted in JS as something like this
function myExtn() {
  // ...
}
```

### Extension method usage 

- extension method must be in scope

```typescript
// proposal
target.myExtn();

// treated the same as typescript
myExtn.bind(target)();

// which emitted in JS as something like this
myExtn.bind(target)();
```

So what's emitted is not based on the types of the expressions. It's just a simple re-arrangement.

Obviously there would be some fancy type checking stuff that I won't try to pretend to understand.

**Comment by MartinJohns** [comment]
You're emitting different JS.

**Comment by fatcerberus** [comment]
It's still type-directed emit because when TS sees the function call expression `x.foo()`, whether that should be emitted as `x.foo()` or `foo.call(x)` in your proposal depends specifically on the type of `x`.  TS will never do this.

**Comment by gregbacchus** [comment]
Oh ok. Thanks for the explanation. I see that now. 

I had imagined that if you imported `foo`, you would always transform. If you don't want the extension, don't import. That makes it very straight forward.

**Comment by FilipeBeck** [comment]
I liked this suggestion. I think the philosophy of language shouldn't be that uncompromising. We already have `enum` and `namespace` that are outside the rule.

**Comment by FilipeBeck** [comment]
I have a lib that extends native types. This feature would fit like a glove.

**Comment by MartinJohns** [comment]
> We already have `enum` and `namespace` that are outside the rule.

At least the addition of `enum` was called a mistake by the team already. And the philosophy exists to protect TypeScript and allow to adopt new ECMAScript features without having to worry about incompatibilities, and to prevent feature creep.

**Comment by fatcerberus** [comment]
> I had imagined that if you imported foo, you would always transform.

Which is *still* type-directed emit because it relies on the compiler knowing the type of `foo`, specifically whether it's an extension method or not, information that doesn't exist at runtime.

This is just something that's outside the current design philosophy of the language, which is to be, essentially, "JavaScript with compile-time type checking".  With the exception of grandfathered features like `enum` and `namespace`, what you write in TS is exactly what runs at runtime in the JS engine, after simply stripping away the types.

Note also that lots of workflows use `tsc` only for type-checking and do the actual transpilation with Babel, so there's a practical reason to not to things like this too: adding type-directed emit features _now_ would break Babel workflows, as Babel just strips the types and (to my knowledge) doesn't implement any of TS's type system at all.
